### ISSUE TITLE
Out of date comments in System.Numerics.Vectors

### ISSUE THREAD

**Issue by fiigii** [dotnet/corefx#15328]
There is a comment at the beginning of  Vectors.cs [https://github.com/dotnet/corefx/blob/master/src/System.Numerics.Vectors/src/System/Numerics/Vector.cs#L22-L32](url)

>  PATTERN:
    *    if (Vector.IsHardwareAccelerated) { ... }
    *    else { ... }
    * EXPLANATION
    *    This pattern solves two problems:
    *        1. Allows us to unroll loops when we know the size (when no hardware acceleration is present)
    *        2. Allows reflection to work:
    *            - If a method is called via reflection, it will not be "intrinsified", which would cause issues if we did
    *              not provide an implementation for that case (i.e. if it only included a case which assumed 16-byte registers)
    *    (NOTE: It is assumed that Vector.IsHardwareAccelerated will be a compile-time constant, eliminating these checks
    *        from the JIT'd code.)

This means that the Vector.IsHardwareAccelerated true path of each SIMD intrinsic fallback would execute in reflection calls.

However, according to my experiment,  CoreCLR’s behavior is different from the comment. The Vector.IsHardwareAccelerated true path cannot execute, even though it called by reflection. Usually, reflection calls of SIMD intrinsics also go to jited intrinsic.

I detected that the true path only executes when the current base type is not supported by current SIMD intrinsic (in CoreCLR). For example, there is a SIMD intrinsic “DotProduct” in CoreCLR, which just supports Int, Long, Float, and Double. If we pass two vectors of Byte or Short, DotProduct would go to the C# fallback (the Vector.IsHardwareAccelerated true path).

Consequently, here are two issues:
1. The comment is different from code's behavior (out of date comments?).
2. For most of the SIMD intrinsics, the code pattern "if (Vector.IsHardwareAccelerated) { ... }" looks useless.

**Comment by mellinoe** [comment]
I believe what you've said about reflection is true, @sivarv do you recall whether this behavior changed at some point, or has the comment always been inaccurate? I know we added these safety codepaths because we were specifically concerned about methods called through reflection being treated differently.

> the true path only executes when the current base type is not supported by current SIMD intrinsic (in CoreCLR). For example, there is a SIMD intrinsic “DotProduct” in CoreCLR, which just supports Int, Long, Float, and Double. If we pass two vectors of Byte or Short, DotProduct would go to the C# fallback (the Vector.IsHardwareAccelerated true path).

This is a good point regarding the current x64 support. On the other hand, we are striving to support a lot of different architectures, with x86 being worked on right now. With more and more architectures, I would think that it will be hard to determine exactly which intrinsics will fall into this category. I'm not as familiar with ARM NEON, for example, but I wouldn't be surprised if different combinations of operations/data-types are supported when compared to our x64 backend, or compared with our x86 backend. We should definitely update the comment, but perhaps the "safety codepaths" are still valuable given the above.

**Comment by fiigii** [comment]
Thank you for the reply @mellinoe .Originally,  I just wanted to make sure the behavior of SIMD intrinsics in reflection calls because my experiments maybe do not cover all situations. Your answer is very clear to me.

Additionally,  I recommend to update this comment with more details about general fallback implementation patterns. 
For example, when are the unrolled fallbacks called? when do the  "safety codepaths" maybe execute? On a machine that does not support SSE2, CoreCLR/CoreFX always treats Vector<T> as 128-bit length, etc.
I think this will be very helpful to developers who attempt to add new SIMD intrinsics into CoreCLR/CoreFX.

**Comment by sivarv** [comment]
@mellinoe @fiigii - I am not sure that I understand why it the following pattern is useless to guard against reflection invocation of a SIMD method.

```
if (Vector.IsHardwareAccelerated) { ... }
 else { ... }

```

Say DotProduct method is called on Vector<float> through reflection.  DotProduct method IL will get jitted, because in the caller method JIT will not recognize a method call via reflection as an intrinsic. Since that method IL has the above if-check, JIT will see Vector.IsHardwareAccelerated is true and will execute the code block associated if-stmnt and that logic is written such it it will work correctly both on SSE2 and AVX2 targets.

In summary, I think the original comment is still correct.

**Comment by fiigii** [comment]
@sivarv @mellinoe Sorry, there was something wrong in my experiments. I ran them again, and I detected that the method body of DotProduct indeed can by called in reflection. However, other SIMD intrinsics are not. Here is my experiment steps:
1. Pick up a SIMD intrinsic to change its fallback in Vector.cs. Here I selected "Abs", and changed its "safety codepath" to return Zero vector always. So if its method is called, the result would be Zero.
```csharp
if (Vector.IsHardwareAccelerated)
{
    if (typeof(T) == typeof(SByte))
    {
        SByte* dataPtr = stackalloc SByte[Count];
        for (int g = 0; g < Count; g++)
        {
            dataPtr[g] = (SByte)(object)(Math.Abs((SByte)(object)value[g]));
         }
         return Vector<T>.Zero;
    }
    else if (typeof(T) == typeof(Int16))
    {
        Int16* dataPtr = stackalloc Int16[Count];
        for (int g = 0; g < Count; g++)
        {
            dataPtr[g] = (Int16)(object)(Math.Abs((Int16)(object)value[g]));
         }
         return Vector<T>.Zero;
    }
    ......
    ......
}
```
2. Build and install the modified CoreFX. Then call the "Abs" intrinsic twice, directly and reflection.
```csharp
var a = new Vector<float>(-2F);
Console.WriteLine(Vector.Abs(a)); //directly call to the intrinsic that should print <2, 2, 2, 2, ....>
        
 var arg = new Object[] {a};
var vectorClass = Assembly.GetAssembly(a.GetType()).GetTypes().Where( x => x.Namespace.Contains("System.Numerics"));
foreach (var klass in vectorClass) {
    if (klass.Name == "Vector")
    {        
        var meth = klass.GetMethod("Abs").MakeGenericMethod(typeof(float));
        Console.WriteLine(meth.Invoke(null, arg)); 
        // reflection call into the fallback that should print <0, 0, 0, 0, .....> 
    }
}
```
However, the output is:
```
<2, 2, 2, 2, 2, 2, 2, 2>
<2, 2, 2, 2, 2, 2, 2, 2>
```
That means the reflection call also gets jited as intrinsic.
I also tried this code for other SIMD methods (i.e. Add, Subtract, Min, Max, etc), and they all call jited intrinsics rather than fallback method bodies.

My computer is Windows 10 and Intel Core i7 6700k (Skylake).
CoreCLR building's version is 1.2.0-beta-24906-0.

In conclusion, the original comment is incorrect on most SIMD methods, but it is correct on DotProduct.
